File: trilobytelib/src/main/java/com/st/trilobyte/models/Flow.java
package com.st.trilobyte.models;

import java.io.File;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

public class Flow implements Serializable {

    public final static int FLOW_VERSION = 1;

    private int version = FLOW_VERSION;

    private transient File file;

    private String id;

    private String category;

    private List<String> board_compatibility = new ArrayList<>();

    private String description;

    private String notes;

    private List<Sensor> sensors = new ArrayList<>();

    private List<Function> functions = new ArrayList<>();

    private List<Flow> flows = new ArrayList<>();

    private List<Output> outputs = new ArrayList<>();

    public Flow() {
        generateId();
    }

    public String getId() {
        return id;
    }

    public void generateId() {
        id = UUID.randomUUID().toString();
    }

    public int getVersion() {
        return version;
    }

    public String getCategory() { return category; }

    public List<String> getBoard_compatibility() { return board_compatibility; }

    public void setBoard_compatibility(String board){
        board_compatibility.add(board);
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(final String description) {
        this.description = description;
    }

    public String getNotes() {
        return notes;
    }

    public void setNotes(final String notes) {
        this.notes = notes;
    }

    public File getFile() {
        return file;
    }

    public void setFile(final File file) {
        this.file = file;
    }

    public List<Sensor> getSensors() {
        return sensors;
    }

    public void setSensors(final List<Sensor> sensors) {
        this.sensors = sensors;
    }

    public List<Function> getFunctions() {
        return functions;
    }

    public List<Flow> getFlows() {
        return flows;
    }

    public void setFlows(final List<Flow> flows) {
        this.flows = flows;
    }

    public void setFunctions(final List<Function> functions) {
        this.functions = functions;
    }

    public List<Output> getOutputs() {
        return outputs;
    }

    public void setOutputs(final List<Output> outputs) {
        this.outputs = outputs;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        final Flow flow = (Flow) o;
        return Objects.equals(id, flow.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    public boolean canBeUploaded() {
        if (getOutputs() == null || getOutputs().size() == 0) {
            return false;
        }
        for (Output output : getOutputs()) {
            if (output.getCanUpload()) {
                return true;
            }
        }
        return false;
    }

    public boolean canBeUsedAsInput() {
        if (getOutputs() == null || getOutputs().size() == 0) {
            return false;
        }
        for (Output output : getOutputs()) {
            if (output.getId().equals(Output.OUTPUT_AS_INPUT_ID)) {
                return true;
            }
        }
        return false;
    }

    @Override
    public String toString() {
        return "Flow{" +
                "version=" + version +
                ", file=" + file +
                ", id='" + id + '\'' +
                ", board_compatibility="+ board_compatibility +
                ", description='" + description + '\'' +
                ", notes='" + notes + '\'' +
                ", sensors=" + sensors +
                ", functions=" + functions +
                ", flows=" + flows +
                ", outputs=" + outputs +
                '}';
    }
}

File: st_flow_demo/src/main/java/com/st/flow_demo/models/Flow.kt
package com.st.flow_demo.models

import androidx.documentfile.provider.DocumentFile
import com.st.blue_sdk.board_catalog.models.Sensor
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.Objects
import java.util.UUID

@Serializable
data class Flow(
    @SerialName(value = "version")
    val version: Int = FLOW_VERSION,
    @SerialName(value = "id")
    var id: String,
    @SerialName(value = "ex_app")
    var ex_app: Int = FLOW_CUSTOM,
    @SerialName(value = "category")
    var category: String? = null,
    @SerialName(value = "expression")
    var expression: Flow? = null,
    @SerialName(value = "statements")
    var statements: List<Flow> = ArrayList(),
    @SerialName(value = "board_compatibility")
    var board_compatibility: ArrayList<String> = ArrayList(),
    @SerialName(value = "description")
    var description: String,
    @SerialName(value = "notes")
    var notes: String? = null,
    @SerialName(value = "sensors")
    var sensors: List<Sensor> = ArrayList(),
    @SerialName(value = "functions")
    var functions: List<Function> = ArrayList(),
    @SerialName(value = "flows")
    var flows: List<Flow> = ArrayList(),
    @SerialName(value = "outputs")
    var outputs: List<Output> = ArrayList()

) {
    @kotlinx.serialization.Transient
    //var file: File? = null
    var file: DocumentFile? = null

    companion object {
        var FLOW_VERSION = 1
        var FLOW_CUSTOM  = 0
    }

    fun generateId() {
        id = UUID.randomUUID().toString()
    }

    fun setBoard_compatibility(board: String) {
        board_compatibility.add(board)
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val flow: Flow = o as Flow
        return id == flow.id
    }

    override fun hashCode(): Int {
        return Objects.hash(id)
    }

    fun canBeUploaded(): Boolean {
        if (outputs.isEmpty()) {
            return false
        }
        for (output in outputs) {
            if (output.canUpload) {
                return true
            }
        }
        return false
    }

    fun canBeUsedAsInput(): Boolean {
        if (outputs.isEmpty()) {
            return false
        }
        for (output in outputs) {
            if (output.id == Output.OUTPUT_AS_INPUT_ID) {
                return true
            }
        }
        return false
    }

    override fun toString(): String {
        return "Flow{" +
                "version=" + version +
                ", file=" + file +
                ", id='" + id + '\'' +
                ", board_compatibility=" + board_compatibility +
                ", description='" + description + '\'' +
                ", notes='" + notes + '\'' +
                ", sensors=" + sensors +
                ", functions=" + functions +
                ", flows=" + flows +
                ", outputs=" + outputs +
                ", ex_app=" + ex_app +
                '}'
    }
}

fun createNewFlow(boardName: String): Flow {
    val flow = Flow(
        version = Flow.FLOW_VERSION,
        id = "",
        category = null,
        expression = null,
        statements = ArrayList(),
        board_compatibility = arrayListOf(boardName),
        description = "New Flow",
        notes = null,
        sensors = ArrayList(),
        functions = ArrayList(),
        flows = ArrayList(),
        outputs = ArrayList()
    )

    flow.generateId()
    return flow
}
